### DISCUSSION TITLE
Form class Rule::unique('db','field')->ignore($this->model->id)  Error ...must not be accessed before initialization

### DISCUSSION THREAD

**Discussion by imapilot** [livewire/livewire#6058] [7. Bugs (archived)]
### Livewire version

V3.0 beta

### Which PHP version are you using?

PHP 8.1

### Steps To Reproduce

Hello,

Loving the Form class and taking that logic out from the component! Im just struggling to get the following working in the new Form class.

Trying to use the following Rule. 
 `Rule::unique('teams','name')->ignore($this->team->id)`

Receive the error: `Typed property App\Livewire\Forms\TeamForm::$team must not be accessed before initialization`

I have tried it both the following situation:
1. `#Rule()` attribute
2. `public function rules()` method

Sorry if im doing something dumb, read through all the docs and goggled as much as i could but nothing obvious jumped out.

**Comment by joshhanley** [comment]
@imapilot thanks for reporting! It seems you haven't initialised your `$team` property on your form, hence the error. So try making sure you have assigned a model to your `$team` property first before running that validation rule.

Hope this helps!

**Comment by imapilot** [comment]
Thankyou for replying!!

```
class TeamForm extends Form
{

    public ?Team $team;
    ...

   public function setPost(Team $team)
   {
          $this->team = $team;
          ...
   }
}
```

Do you mean the `public ?Team $team;` and setting it in `setPost`? I believe I have them set, I think I'm having trouble with using the Livewire\Attributes\Rule attribute with Illuminate\Validation\Rule::unique.

I was referencing the following parts of the docs https://livewire.laravel.com/docs/validation#custom-key  https://livewire.laravel.com/docs/validation#custom-key


Below gives **Constant expression contains invalid operations**
```
<?php

namespace App\Livewire\Forms;

use Livewire\Form;
use App\Models\Team;
use Illuminate\Validation\Rule as ValidationRule;
use Livewire\Attributes\Rule;

class TeamForm extends Form
{

    public ?Team $team;

    #[Rule([
        'required',
        ValidationRule::unique('teams','name')->ignore($this->team->id)
    ])]
    public $name = '';

    public function setPost(Team $team)
    {
        $this->team = $team;
        $this->name = $team->name;
    }

    public function store()
    {
        Team::create($this->all());
    }

    public function update()
    {
        $this->team->update(
            $this->all()
        );
    }

}
```

Using the method rules() - this is the way I normally do it in livewire 2 within the component without issue.
Below gives **Typed property App\Livewire\Forms\TeamForm::$team must not be accessed before initialization**
```
<?php

namespace App\Livewire\Forms;

use Livewire\Form;
use App\Models\Team;
use Illuminate\Validation\Rule;
// use Livewire\Attributes\Rule;

class TeamForm extends Form
{

    public ?Team $team;

    public $name = '';

    public function setPost(Team $team)
    {
        $this->team = $team;
        $this->name = $team->name;
    }

    public function rules()
    {
        return [
            'name' => [
                'required',
                Rule::unique('teams','name')->ignore($this->team->id)
            ],
        ];
    }

    public function store()
    {
        Team::create($this->all());
    }

    public function update()
    {
        $this->team->update(
            $this->all()
        );
    }

}
```

Please let me know the correct way to do this, love your work!

**Reply by m15studio** [reply]
I have same problem.
You could declare public $id_team. 
In setPost you could set $this->id_team=$team->id
but validation rule will be completely transparent in form validate

**Reply by m15studio** [reply]
I resolved just now moving rules() on Component Object instead Form Object.
In Component Object it is necessary to do mount() of $team
After this, in the Rule, ignore() seems work it

But the question is:
It's the correct way for custom validation?

**Comment by imapilot** [comment]
I tried the assign the $id_team idea, with rules() but it wasnt working, _i think_ the $id_team was null when the rules are created and therefore the ->ignore() part didnt work. Also tried with manually calling $this->validate() but no luck. Going back to the comonent doesnt seem right?!?


I could get it working by creating a custom Rule that is the same as the Rule::unique but this doesnt '_feel_' like this would be the solution.


```
<?php

namespace App\Livewire\Forms;

use Livewire\Form;
use App\Models\Team;
use Livewire\Attributes\Rule;
use App\Rules\UniqueFieldForTeam;

class TeamForm extends Form
{

    public ?Team $team;

    #[Rule([
        'required',
        new UniqueFieldForTeam

    ])]
    public $name = '';

    public function setPost(Team $team)
    {
        $this->team = $team;
        $this->name = $team->name;
    }

    public function store()
    {
        Team::create($this->all());
    }

    public function update()
    {
        // Had to remove the team property otherwise:
        //      Add fillable property [team] to allow mass assignment on [App\Models\Team]
        // [docs seem to show no issues??]
        $this->team->update(
            collect($this->all())->except(['team'])->toArray()
        );
    }

}
```


```
<?php

namespace App\Rules;

use Closure;
use App\Models\Team;
use Illuminate\Support\Str;
use Illuminate\Contracts\Validation\DataAwareRule;
use Illuminate\Contracts\Validation\ValidationRule;

class UniqueFieldForTeam implements DataAwareRule, ValidationRule
{
    protected $data = [];

    public function validate(string $attribute, mixed $value, Closure $fail): void
    {
        $attribute = Str::remove('form.', $attribute);
        if (Team::where($attribute, $value)
                ->where('id', '!=', $this->data['team']->id)
                ->exists()) {
            $fail('The :attribute already exists');
        }
    }

    public function setData($data)
    {
        $this->data = $data['form'];

        return $this;
    }
}
```

**Comment by m15studio** [comment]
You must move Rule from Form Object to Component Object.
After doing this in the Component you have $id_team valorized and rule works

**Comment by imapilot** [comment]
Isn't the point of the Form object to remove all of the Form related stuff out of the Component? Seems like a step backwards?

**Comment by DanielSpravtsev** [comment]
@imapilot @m15studio 
You can use #Rule attributes with rules() public function. For example:

```
class ServerForm extends Form
{
    #[Rule('string|nullable')]
    public ?string $id = '';

    #[Rule(['required', 'min:3', 'ip'])]
    public ?string $ip;
    
    public function rules()
    {
        return [
            'ip' => 'unique:servers,ip,'.$this->id // don't use Rule facade or Rule functions, they will not work, only text rules. OR add ->__toString() at the end
        ];
    }
    
}
```

Rules from attributes and public method will be merged

**Reply by imapilot** [reply]
@DanielSpravtsev  _// don't use Rule facade or Rule functions, they will not work, only text rules. OR add ->__toString() at the end_

Is this in the rules() method or in general, ie in Form Object?

Losing the Rule facade is disappointing, losing the ability to add a where() or closure within the validation without creating a whole custom rule was convinent!
Can you show me an example of the __toString()? that one flew passed me.

Thankyou.

**Reply by DanielSpravtsev** [reply]
In the current beta, Rule facade don't work in Form Objects
One way to make it work — manually convert it toString
For example: 
```
public function rules()
    {
        return [
            'form.some_field' => \Illuminate\Validation\Rule::unique('table', 'column')->ignore($this->id)->__toString(),
        ];
    }
```

**Reply by m15studio** [reply]
@DanielSpravtsev I've yet problems with mount

```
class ClienteForm extends Form {
 public  ? Cliente $cliente;
    #[Locked]
public $id_cliente;

public function setCliente(Cliente $cliente) {
    $this->cliente = $cliente;
    $this->id_cliente = $cliente->id_cliente;
}

public function rules() {
    return [
        'cliente_email' => ['required', 'email', 'max:255', \Illuminate\Validation\Rule::unique('cliente', 'cliente_email')->ignore($this->id_cliente, 'id_cliente')->__toString()],
];
}
```

`$this->id_cliente` is NULL

this is a debug extract
```
"clienteForm.cliente_email" => array:4 [
            0 => "required"
            1 => "email"
            2 => "max:255"
            3 => "unique:cliente,cliente_email,NULL,id_cliente"
          ]
```

`setCliente` function is called by `mount()` function of Cliente class (component of form)

**Reply by DanielSpravtsev** [reply]
I think it's a bug https://github.com/livewire/livewire/discussions/6174

**Comment by m15studio** [comment]
@DanielSpravtsev 
I also tested your tip before find my temp solution.
You are right, rules will be merged, but the problem is that $this->id is null

**Reply by DanielSpravtsev** [reply]
You can mount it yourself. 

If you using form object, you can create  something like 'setModel' or 'fill' function at formObject class and then set whatever you need with component mount method. 


For example 

public function mount (Post $post){
!$post ?: $this->postForm->setModel($post);
}


It'll be much easier to help, if you share your component and form classes

**Reply by m15studio** [reply]
@DanielSpravtsev thnks a lot
with setModel function custom rule works.

But, custom messages e custom validationAttribut must to be in Post, because in PostForm they don't work.

In blade I have eg.
`<input text wire:model="postform.title"/>`

and if I put custom messages in PostForm they must be referral to "title" property of PostForm and not for postform.title 

another little question: 
If I put custom Rule in Post it overwrites same Rule in PostForm?
eg.
I have NewPost and EditPost with common PostForm
For NewPost I have a certain rule, but same rule for EditPost is different

**Reply by joshhanley** [reply]
@m15studio There is currently an open PR to add messages to FormObjects, see https://github.com/livewire/livewire/pull/5904

**Comment by joshhanley** [comment]
@imapilot someone raised this discussion on Caleb's stream. I just tested this, and the issue is to do with the uninitialised `$team` property.

I was able to replicate the issue, and the fix is to assign team to `null` initially `public ?Team $team = null;` then it should all work happily.

And yeah because you're referencing a validation object (unique), you need to use the `rules()` method instead of the `#[Rule]` attribute.

Hope this helps!
